Automattic/woocommerce-payments
# Basic Support for Analytics

Fixes #2249 

#### Changes proposed in this Pull Request









* Adds analytics for the Multi-Currency plugin.
* Currently, this will just be making sure any purchases in a non-store currency are displayed as part of the table/reports (converted into the store default currency using the exchange rate at the time of purchase).

#### Notes on functionality

- All *multi-currency* order/product/tax/coupon values are converted to the default store currency on all standard reporting types, and all composite values which appear on these pages should also be calculated using the converted values where applicable (e.g. Net sales will be the sum of all of the net total value of all orders, converted where required). Any custom, extensible report types added by other plugins _should_ use the converted values because we have logic to do the conversions on a generic report type.
- A "Customer Currency" field has been added to the `Orders` tables (which should take effect wherever these are used in WC Admin). It was not possible to add this field to the other standard report types (Products, Revenue, Orders, Coupons, Taxes), because these report types show grouped values. For example, in the case of products, it is grouped by product type, so it's not possible to display one 'customer currency' for a product type, since there could have been many orders of that product made using different currencies. Having said that, all *values* which are used in the calculation of the amount displayed in the report will be converted correctly.
- There is a known issue where the "Leaderboards" table on the *Overview* report does *not* respect the store currency. It seems to display in the currency of the first order in the list, which is a known bug with WooCommerce Admin: https://github.com/woocommerce/woocommerce-admin/issues/5154. It seems this issue has not yet been fixed, so I think we should liase with them to see if it is possible to put a fix in place for this. It looks like Team Mothra could potentially assist us with this.
- Multi-Currency orders that were made before the latest update to this branch was present in your branch will not display properly in the analytics, because the conversion would not have been made when the order data was saved to the DB.

#### Testing instructions





Note that I am still in the process of adding unit tests, but code reviews and manual testing would be very welcome. :)

* Please start by taking a look at the Notes on Functionality above as there is one known issue which is related to WC Admin. :)
* Make a few orders in your store using multi-currency in a few different currencies, and a few using the store default currency (to test that these amounts do *not* get converted). In order to test all areas of the analytics, I would recommend ordering the following:
    - A few different products
    - Belonging to different product categories
    - Some orders using discount coupons and some without
    - Some variations of the same product
    - Some orders with taxes and shipping applied.
* Go to `Tools->Scheduled Actions` in your WP Admin, and run any `wc-admin_import_orders` jobs that you see relating to the orders you just made.
* Go to the different sections of the analytics reporting (`Overview`, `Products`, `Revenue`, `Orders`, `Variations`, `Categories`, `Coupons`, `Taxes`, `Downloads`, `Stock`) and look at the report results. Verify that the results look sensible, that they contain the orders you have made, and importantly the amounts should be all converted into your store's default currency. In the case of orders made using the store currency, they should just use the order amount. In the case of multi-currency orders, they should use the order amount, converted back into the store currency using the exchange rate at the time of transaction.
* Check the other areas of the analytics interface to make sure there are no errors and the data makes sense This could include: changing the report type/report filters, changing the chart type/chart display options, searching the table, downloading a CSV, hiding/showing columns on the table).

-------------------
